File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.Arrays;
import java.util.Set;
import java.util.HashSet;

class Solution {
    public boolean solution(String[] phone_book) {
        boolean answer = true;
        
        Arrays.sort(phone_book, (a, b) -> a.length() - b.length());
        Set<String> set = new HashSet<>();
        
        outer: for (String phoneNumber : phone_book){
            for (int i = 0; i < phoneNumber.length(); ++i){
                if(set.contains(phoneNumber.substring(0,i))){
                    answer = false;
                    break outer;
                }
                
                set.add(phoneNumber);
            }
        }

        
        return answer;
    }
}

File: 백준/Gold/5052. 전화번호 목록/전화번호 목록.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace ConsoleApp1
{
    internal class Program
    {
        public static void Main(string[] args)
        {
            StreamReader sr = new StreamReader(new BufferedStream(Console.OpenStandardInput()));
            StreamWriter sw = new StreamWriter(new BufferedStream(Console.OpenStandardOutput()));

            int T = Int32.Parse(sr.ReadLine());

            StringBuilder answer = new StringBuilder();
            for (int i = 0; i < T; i++)
            {
                int N = Int32.Parse(sr.ReadLine());

                List<string> list = new List<string>();
                for (int j = 0; j < N; j++)
                {
                    string curPhoneNumber = sr.ReadLine();
                    list.Add(curPhoneNumber);
                }

                list.Sort((a, b) => a.Length - b.Length);

                HashSet<string> set = new HashSet<string>(list);

                bool check = true;
                foreach(var num in list)
                {
                    if (!check) break;

                    string strNum = num.ToString();

                    for (int j = 1; j < strNum.Length; j++)
                    {
                        if (set.Contains(strNum.Substring(0, j)))
                        {
                            check = false;
                            break;
                        }
                    }
                }

                if (check)
                {
                    answer.Append("YES\n");
                    continue;
                }
                answer.Append("NO\n");
            }

            sw.Write(answer.ToString());
            
            sw.Close();
            sr.Close();
        }
    }
}
